Merged the IR and Vis uncertainty changes

mraspaud opened the pull request:
into code including a new module

which merges the two uncertainty outputs into a single xarray output.

Commit:
Added vis_uncertainty

Commit:
Added vis_uncertainty

Commit:
Added systematic uncertainty components

Commit:
tidying up

Commit:
added function get_bad_space_counts to utils

Commit:
fixed solar contamination flag

Commit:
added command line test for visible uncertainty

Commit:
added flag for bad space counts

Commit:
Preparing code for v1 delivery

jmittaz committed:
Merged the IR and Vis uncertainty changes into code including a new module
which merges the two uncertainty outputs into a single xarray output.

nicoleyaghnam committed:
Update vis_uncertainty.py

Fixed issue causing NaNs in uncertainty output

Commit:
Merge remote-tracking branch 'origin/combined_uncertainties' into visible

# Conflicts:
#	pygac/calibration/vis_uncertainty.py
#	pygac/klm_reader.py
#	pygac/pod_reader.py
#	pygac/reader.py
#	pygac/utils.py

nicoleyaghnam committed:
Merge pull request #1 from NPL-EO/visible

Visible

Commit:
Fixed issue of NaNs appearing in uncertainty output

Commit:
Fixed issue of logical chan_3a overriding other channels in gain evaluation

jmittaz committed:
Modified version of code to remove some of the errors/warnings by pygac
git repository

jmittaz committed:
Merge pull request #2 from NPL-EO/visible

Fixed issue of NaNs appearing in uncertainty output

jmittaz committed:
Fix output dataset through calibration to include the raw counts rather than
calibrated values ("counts" rather than "channels")

jmittaz committed:
Fixed potential bug in iprt numbering routine (line_diff outside 1-4)

jmittaz committed:
Moved "counts" xarray DataArray to allow for different calibrators

jmittaz committed:
Update solar contamination routine to allow for detection at two locations
(before/after solZA max)

jmittaz committed:
Fix a bug in flagging plus extra plotting

jmittaz committed:
Fixed a few bugs related to plotting and added tests for uncertainty ratio
out of limits

jmittaz committed:
Changed systematic uncertainty for 3.7 micron channel to allow for Planck
function variations but adding 0.5K@300K in radiance space for this component

jmittaz committed:
Changed 3.7mu sys uncertainty to take Planck variation into account (uses
radiance at 300K assuming uncertainty of 0.5K) and updates to data files
with new solar contamination routine.

jmittaz committed:
Added to possible outputs for solar contamination detection and added a
write_solar option to output data to be used in unit testing

jmittaz committed:
Added tests for updated noaa calibration module and for IR uncertainty.
	1) Added test for updated PRT numbering module
	2) Added test for aspects of the IR uncertainty module including
	solar contamination of the gain, calculation of the min PRT standard
	deviation gain and calculation of the ICT uncertainty. Uses data from
	an orbit of real AVHRR data stored in associated CSV file.

jmittaz committed:
Modified csv_file to correct location

jmittaz committed:
Added test for noise calculation of IR channels

jmittaz committed:
Modified code to change default command line procedure to not have plotting
for uncertainy.py and vis_uncertainty.py

mraspaud commented:
Dear Jon,

The test are run with the pytest command, which will try to find and execute all functions and methods which name starts with `test` I think.

In principle, if you have pytest install, you should be able to run `pytest .` from the pygac directory to check that the tests are passing. So by uninstalling matplotlib first, you should be able to reproduce the errors we have here:
https://github.com/pytroll/pygac/actions/runs/15190080899/job/42768625974?pr=144

eg

```
__________________ ERROR collecting pygac/tests/test_utils.py __________________
ImportError while importing test module '/home/runner/work/pygac/pygac/pygac/tests/test_utils.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/share/miniconda3/envs/test-environment/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
pygac/__init__.py:35: in <module>
    from pygac.runner import get_reader_class, process_file  # noqa
pygac/runner.py:35: in <module>
    from pygac.gac_klm import GACKLMReader
pygac/gac_klm.py:33: in <module>
    from pygac.gac_reader import GACReader
pygac/gac_reader.py:44: in <module>
    from pygac.reader import Reader, ReaderError
pygac/reader.py:46: in <module>
    from pygac import gac_io
pygac/gac_io.py:36: in <module>
    from pygac.utils import check_user_scanlines, slice_channel, strip_invalid_lat
pygac/utils.py:28: in <module>
    import matplotlib.pyplot as plt
E   ModuleNotFoundError: No module named 'matplotlib'

```

jmittaz committed:
Removed top level import of matplotlib

mraspaud commented:
Dear Jon,
Looks like there is another on in pod_reader.py...

jmittaz committed:
Remove reference to matplotlib

mraspaud commented:
I'll let you check the last errors here: https://github.com/pytroll/pygac/actions/runs/15208648467/job/42780018707?pr=144

jmittaz committed:
Fix bugs in test routine

jmittaz commented:
Dear Martin,

Fixed the bug case but the “import cftime” is not a bug but a python environment thing so what do you suggest? I also use cftime in “ir_uncertainty.py”

Jon.

Dr Jonathan Mittaz
Room 3L64
Dept. of Meteorology
Whiteknights Road
Earley Gate
Reading
RG6 6ET
Telephone: 0118 3786732

If your email is a request under the Environmental Information Regulations or Freedom of Information then please resend this to ***@***.***




From: Martin Raspaud ***@***.***>
Date: Friday, 23 May 2025 at 13:02
To: pytroll/pygac ***@***.***>
Cc: Jonathan Mittaz ***@***.***>, Mention ***@***.***>
Subject: Re: [pytroll/pygac] Merged the IR and Vis uncertainty changes (PR #144)
[Image removed by sender.]mraspaud left a comment (pytroll/pygac#144)<https://github.com/pytroll/pygac/pull/144#issuecomment-2904197819>

I'll let you check the last errors here: https://github.com/pytroll/pygac/actions/runs/15208648467/job/42780018707?pr=144

—
Reply to this email directly, view it on GitHub<https://github.com/pytroll/pygac/pull/144#issuecomment-2904197819>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABTPQXHR3TIPGVCTJOQZDBL274E3HAVCNFSM6AAAAAB26P7XRCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMBUGE4TOOBRHE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

mraspaud commented:
I can see indeed that cftime is imported in ir_uncertainty.py, but it doesn't seem to be used anywhere in that file? I can see it is used though in the tests though. If it is used just for tests, it just needs to be added to the CI, here https://github.com/pytroll/pygac/blob/main/continuous_integration/environment.yaml

jmittaz committed:
Removed *_uncert.nc data which is now stored in zenodo. Added test for zenodo
files in test routine and made changes to ir_uncertainty.py to use zenodo files

jmittaz committed:
Added cftime to environment

jmittaz committed:
Added test for zenodo file opening to trap no file available

mraspaud commented:
@jmittaz As we have merged a few things to the main branch, we need to resolve some conflicts with you branch here. I don't have permissions to do it, would mind allowing maintainers to add commits here so I can help? I think it's probably a setting on this page, to the right I suppose.

mraspaud commented on pygac/runner.py:
```diff
@@ -43,7 +43,7 @@ class for a given file.
 
 
-def get_reader_class(filename, fileobj=None):
+def  get_reader_class(filename, fileobj=None):
```
```suggestion
def get_reader_class(filename, fileobj=None):
```

mraspaud commented on pygac/klm_reader.py:
```diff
@@ -734,14 +734,20 @@ def _validate_header(cls, header):
 
     def get_telemetry(self):
-        """Get the telemetry.
+        """Get the telemetry. Modified by J.Mittaz / University of Reading
```
Git blame can provide information of authorship and changes, so please remove the attributions and description of updates from the code.

Also: docstring should have just one summary line, ending with a dot.

jmittaz committed:
Added visible channel test routines and made modifications to klm_reader.py
and noaa.py following requests from Martin Raspaud SHMI

jmittaz committed:
Moved matplotlib include

jmittaz committed:
Added filenotfound exception to try:except loops

mraspaud commented on pygac/calibration/ir_uncertainty.py:
```diff
@@ -1122,7 +1122,7 @@ def get_gainval(time,avhrr,ict1,ict2,ict3,ict4,CS,CICT,CE,NS,bad_scan,convT,
     try:
         d = xr.open_dataset(coef_file)
-    except:
+    except FileNotFoundError::
```
There is one ":" too many here...

mraspaud commented on pygac/calibration/ir_uncertainty.py:
```diff
@@ -1234,7 +1234,7 @@ def get_solar_from_file(avhrr_name,ds):
     try:
         d = xr.open_dataset(coef_file,decode_times=False)
-    except:
+    except FileNotFoundError::
```
There is one ":" too many here...

jmittaz committed:
Fixed bug in except sections

mraspaud commented:
Thanks! Any news about allowing me to help with this PR?

jmittaz commented:
I’ll ask Sam again

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Martin Raspaud ***@***.***>
Sent: Wednesday, June 11, 2025 9:57:22 AM
To: pytroll/pygac ***@***.***>
Cc: Jonathan Mittaz ***@***.***>; Mention ***@***.***>
Subject: Re: [pytroll/pygac] Merged the IR and Vis uncertainty changes (PR #144)

[https://avatars.githubusercontent.com/u/167802?s=20&v=4]mraspaud left a comment (pytroll/pygac#144)<https://github.com/pytroll/pygac/pull/144#issuecomment-2961798385>

Thanks! Any news about allowing me to help with this PR?

—
Reply to this email directly, view it on GitHub<https://github.com/pytroll/pygac/pull/144#issuecomment-2961798385>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABTPQXGURMAKUBFE7RAZ7VL3C7VPFAVCNFSM6AAAAAB26P7XRCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNRRG44TQMZYGU>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

mraspaud committed:
Merge branch 'main' into combined_uncertainties

mraspaud commented:
I have now merged the newest changes to pygac into this branch by resolving the conflicts

mraspaud reviewed (commented):
First part of a review.

Commit:
Updated test_vis_uncertainty.py to account for functions changing location

mraspaud committed:
Merge branch 'main' into combined_uncertainties

Commit:
Added seaborn to requirements.txt

mraspaud commented on requirements.txt:
```diff
@@ -4,3 +4,4 @@ scipy>=0.8.0
 python-geotiepoints>=1.1.8
 packaging>=14.0
 xarray>=2024.1.0
+seaborn>=0.12.2
```
this shouldn't be required for production code.

mraspaud committed:
Remove seaborn from hard dependencies, and clean up some code

Commit:
Added import "files" to test_ir_uncertainty.py

jmittaz committed:
Add files from importlib.resources

jmittaz committed:
Fixed bugs

jmittaz committed:
Added in extra test to test complete IR uncertainty package. Also added
data to associated csv file

jmittaz committed:
Merge branch 'combined_uncertainties' of https://github.com/NPL-EO/pygac into combined_uncertainties

jmittaz committed:
Fix bug with test data filename

jmittaz committed:
Fix bug in Uict test

jmittaz committed:
Fix error in get_telemetry in klm_reader.py and fix associated get_telemetry
setup in test_reader.py

jmittaz committed:
Added netCDF4 to environment

jmittaz committed:
Add get_vis_telemetry to fake POD reader class for testing

nicoleyaghnam committed:
Merge remote-tracking branch 'upstream/main' into combined_uncertainties

nicoleyaghnam committed:
Replaced 'self.reader.tle_lines = ("first line", "second line")' with actual TLE data, unit tests in test_klm.py now passing

nicoleyaghnam committed:
Fixed TLE path

jmittaz committed:
Removed redundant line

jmittaz committed:
Made changes to many tests/test data to correct problems related to the
addition of get_angles in create_counts_data in reader.py

mraspaud committed:
Fix last test

mraspaud committed:
Remove spurious breakpoint

jmittaz committed:
Fixed spelling error and variable name error (avhrr_name)

mraspaud committed:
Cache zenodo files

mraspaud committed:
Add missing dependencies to CI env

mraspaud committed:
Fix minimal python version

mraspaud committed:
Simplify Allan deviation computation

mraspaud committed:
Update pygac/runner.py

jmittaz committed:
Fix small bug in ir_uncertainty and updated associated tests/test data

jmittaz committed:
Merge branch 'combined_uncertainties' of https://github.com/NPL-EO/pygac into combined_uncertainties

mraspaud committed:
Add uncertainty to calibrated dataset

mraspaud committed:
Fix median gcp distance assignment

jmittaz committed:
Added code to support creation of auxilliary data (min gain and solar
contamination)

jmittaz committed:
Merge branch 'combined_uncertainties' of https://github.com/NPL-EO/pygac into combined_uncertainties

jmittaz committed:
Changed zenodo reference to updated file location

jmittaz committed:
Fixed bugs in IR calibration (zenodo naming and dealing with missing 3.7
micron data) Vis calibration (accessing 1.6 micron channel) and output uncertainty (flag changes)

jmittaz committed:
Fixed bug in uratio calculation

jmittaz committed:
Fixed uncertainty flags metadata

mraspaud committed:
Do not crash on failing georeferencing and uncertainties

nicoleyaghnam committed:
Fixed Rcal evaluation to store for all scanlines

mraspaud committed:
Fix time offset correction

nicoleyaghnam committed:
Fixed shape mis-match with Rcal and noise

mraspaud committed:
Switch on cf compatibility for uncertainty flags

mraspaud committed:
Add quality_flags in dataset

mraspaud committed:
Warn when georeferencing or uncertainty computations fail

mraspaud committed:
Fix qual_flags' meanings

nicoleyaghnam committed:
Fixed Rcal evaluation to reduce run time

nicoleyaghnam committed:
Fixed Rcal evaluation to reduce run time

mraspaud committed:
Add estimated time offset and attitude as attributes

mraspaud committed:
Refactor telemetry extraction and access
